store: Handle invalid API key on register-queue

PIG208 opened the pull request:
~~This is near term fix for a user-reported issue:
  https://chat.zulip.org/#narrow/channel/48-mobile/topic/0.2E0.2E19.20Flutter.20.3A.20Cant.20connect.20to.20self.20hosted.20instance/near/2004042~~

~~It is not intended to be the full fix.  With a better UX, we would bring the user back to the choose-account page without them manually doing so.  That's covered by #737 but out-of-scope for this commit.~~

This cherry-picks https://github.com/zulip/zulip-flutter/pull/1235, and partially addresses #890, by handling `INVALID_API_KEY` errors from the `PerAccountStoreWidget`.

Fixes: #737

gnprice commented:
Thanks!

Would you post a screenshot of what the error looks like? Then we can also wordsmith the error message.

PIG208 commented:
Sure!

<details>
<summary>Details</summary>

![image](https://github.com/user-attachments/assets/3c70d53b-281b-4291-b1a8-85510e9f9d47)

</details>

gnprice commented:
Cool, thanks.

I think "Invalid API key" is probably too technical for this context. @alya do we have an appropriate message handy from elsewhere in the product?

PIG208 commented:
Perhaps we can reuse the

PIG208 commented:
Perhaps we can rename and reuse the title string we have for login failure:

```json
  "errorLoginCouldNotConnectTitle": "Could not connect",
```

This sort of feels like an instance of #741 the way we currently handle it, but this message will not be needed after #737.

PIG208 commented:
Updated the PR to implement Chris' suggestion [here](https://chat.zulip.org/#narrow/channel/48-mobile/topic/.23F1183.20Invalid.20API.20key.20response/near/2018994). We now handle invalid API key by calling `logOutAccount`, which removes the routes associated with the affected account, pushing a choose-account page route onto the stack if it becomes empty, and showing an error dialog.

chrisbobbe commented on assets/l10n/app_en.arb:
```diff
@@ -186,8 +186,8 @@
-  "errorLoginCouldNotConnectTitle": "Could not connect",
-  "@errorLoginCouldNotConnectTitle": {
+  "errorCouldNotConnectTitle": "Could not connect",
+  "@errorCouldNotConnectTitle": {
```
```
l10n [nfc]: Use a generalize name for errorCouldNotConnectTitle
```

commit-message nit: "generalized"

chrisbobbe commented on lib/widgets/home.dart:
```diff
@@ -180,9 +185,15 @@ class _LoadingPlaceholderPageState extends State<_LoadingPlaceholderPage> {
-      .getAccount(widget.accountId)!.realmUrl;
+    final account = GlobalStoreWidget.of(context).getAccount(widget.accountId);
+
+    if (account == null) {
+      // We should only reach this state very briefly.
+      // See [_LoadingPlaceholderPage.accountId].
+      return Scaffold(
+        appBar: AppBar(),
+        body: const SizedBox.shrink());
+    }
```
This feels like it duplicates more details than it needs to, like the `appBar` param and the whole set of `Scaffold` params that aren't passed. How about applying the condition more precisely to the things that are supposed to be different?

chrisbobbe commented on lib/log.dart:
```diff
@@ -50,6 +51,17 @@ typedef ReportErrorCancellablyCallback = void Function(String? message, {String?
+// This gets set in [ZulipApp].  We need this indirection to keep `lib/log.dart`
+// from importing widget code, because the file is a dependency for the rest of
+// the app.
+ReportErrorCallback reportErrorToUserModally = defaultReportErrorToUserBriefly;
```
Setting `reportErrorToUserModally` to `defaultReportErrorToUserBriefly` looks odd. How about renaming `defaultReportErrorToUserBriefly` to something that seems like an appropriate value for both `reportErrorToUserBriefly` and `reportErrorToUserModally`?

Maybe `dumpErrorToConsole`? That name makes it sound like a potential problem if it actually gets called (errors should be reported, not dropped on the floor 🙂)—but that's actually appropriate in this case.

PIG208 replied:
Good idea! Renaming to `reportErrorToConsole` because we don't usually dump the error object.

chrisbobbe commented on test/widgets/app_test.dart:
```diff
@@ -361,5 +362,24 @@ void main() {
+      reportErrorToUserModally(message, details: details);
+      check(ZulipApp.ready).value.isFalse();
+      await tester.pump();
+      check(find.byType(AlertDialog)).findsNothing();
```
I have a commit

```
test [nfc]: Factor out checkNoErrorDialog helper
```

in my revision for #1239; how about cherry-picking that as a prep commit and using it here? (See reasoning in the commit)

chrisbobbe reviewed (commented):
Thanks! Here's a review of the first five commits:

06c554338 l10n [nfc]: Use a generalize name for errorCouldNotConnectTitle
d27b46cc2 log [nfc]: Rename ReportErrorCallback to ReportErrorCancellablyCallback
c5fc2d93e example_data: Start generating account id
652c3328e home: Stop assuming account existence from loading page
ab9148088 log: Add reportErrorModally

which leaves the last three that I haven't read yet:

58fd4b160 store: Handle invalid API key on register-queue
7a632b2c7 display test [nfc]: Make openNotification public
2e0bcb944 actions test: Use a realistic example

For the last commit, it's the one with the `Fixes: ` line but it doesn't actually change any app code, it only touches test code. Is that intentional? I'd like to understand if/how these test changes are related to your app-code changes for #737.

PIG208 commented:
The "Fixes" comment gets added to the last commit because it resolves a TODO item for the issue.

chrisbobbe commented on lib/widgets/store.dart:
```diff
@@ -245,11 +261,28 @@ class _PerAccountStoreWidgetState extends State<PerAccountStoreWidget> {
+              // (usually by using [routeToRemoveOnLogout]).
+              return;
+            case ZulipApiException(code: 'INVALID_API_KEY'):
+              // The API key is invalid and the store can never be loaded
+              // unless the user retries manually.
+              if (!mounted) return;
+              final zulipLocalizations = ZulipLocalizations.of(context);
+              reportErrorToUserModally(
+                zulipLocalizations.errorCouldNotConnectTitle,
+                details: zulipLocalizations.errorInvalidApiKeyMessage(
+                  globalStore.getAccount(widget.accountId)!.realmUrl.toString()));
+              unawaited(logOutAccount(context, widget.accountId));
+              return;
```
We also need tests to check that the account is in fact logged out / removed, and that some subtle related things are dealt with, like `_perAccountStores` and `_perAccountStoresLoading`, calling `PerAccountStore.dispose`, and notifying listeners. How about using `TestGlobalStore.takeDoRemoveAccountCalls` for this, because `GlobalStore.removeAccount` already has tests for those things.

chrisbobbe replied:
For example (just for the first-`/register` case):

```dart
  test('GlobalStore.perAccount on INVALID_API_KEY', () => awaitFakeAsync((async) async {
    addTearDown(testBinding.reset);

    await testBinding.globalStore.insertAccount(eg.selfAccount.toCompanion(false));
    testBinding.globalStore.loadPerAccountException = ZulipApiException(
      routeName: '/register', code: 'INVALID_API_KEY', httpStatus: 400,
      data: {}, message: '');
    await check(testBinding.globalStore.perAccount(eg.selfAccount.id))
      .throws<ZulipApiException>();

    // fails
    check(testBinding.globalStore.takeDoRemoveAccountCalls())
      .single.equals(eg.selfAccount.id);
  }));
```

PIG208 replied:
I think the right place for calling `logOutAccount` and `reportErrorModally` should be `loadPerAccount`. We throw an `AccountNotFoundException` after processing the error, and make sure that all call-sites handles that correctly. `_PerAccountStoreWidgetState` already expects the exception.

For the other call-site, we need to handle it at least for one of `_reloadPerAccount`, `UpdateMachine._handlePollError`, or `UpdateMachine.poll`.

It doesn't seem right to catch it from `_reloadPerAccount` without rethrowing or at least indicating that the reload was a failure; `_handlePollError`'s caller assume that the store is disposed after calling it (because of a reloaded `PerAccountStore`), so we should still either rethrow the error or indicate the failure reloading the store. So we probably should just catch it from `poll`, within another `catch` block.

---

Because `logOutAccount` is no longer purely a user action, we might want to revisit https://github.com/zulip/zulip-flutter/pull/1010#discussion_r1817457287 and figure out a new home to this. While `GlobalStore` is logically a good fit, I agree that `lib/model/store.dart` has become too crowded for us to add more to it. Maybe splitting `store.dart` would be a helpful prep/follow-up change.

chrisbobbe commented on test/widgets/store_test.dart:
```diff
@@ -113,6 +119,45 @@ void main() {
+    await tester.tap(find.text('Try another account'));
+    await tester.pump(); // tap the button
+    await tester.pump(const Duration(milliseconds: 250)); // wait for animation
+    check(find.byType(CircularProgressIndicator)).findsOne();
```
I don't yet understand this `CircularProgressIndicator` check.
- Does the test expect it to be on a specific page? (meaning the check stops being as useful if another page adds a `CircularProgressIndicator`)
- If we're expecting it on a non-active/foregrounded page, do we need to use `skipOffstage: false`?

PIG208 replied:
It's looking for the route that is not on the top of the stack. I'm switching to a rewrite of this using `NavigatorObserver`'s, so that we can check the navigator stack more explicitly.

chrisbobbe commented on test/widgets/actions_test.dart:
```diff
@@ -168,57 +169,46 @@ void main() {
-
-      final pushedRoutes = <Route<dynamic>>[];
-      testNavObserver.onPushed = (route, prevRoute) => pushedRoutes.add(route);
-      // TODO(#737): switch to a realistic setup:
-      //   https://github.com/zulip/zulip-flutter/pull/1076#discussion_r1874124363
```
Reading that linked discussion and links from there, I'm not finding an explanation for why the TODO involves #737 at all. What Greg meant by not "realistic setup" is just that "These pages don't exist in real life", meaning two `MaterialAccountWidgetRoute`s with page `InboxPageBody`. Routes like that don't exist in the app because, since the bottom-tabs update, the Inbox page is never alone on its own route, it's always part of the route returned by `HomePage.buildRoute`.

The direct fix is to use realistic routes for `account1Route` and `account2Route`, e.g. by using `MessageListPage.buildRoute` instead, making follow-on changes to `findAccount{1,2}PageContent`, and renaming the `makeUnreadTopicInInbox` helper. Let's have a commit that just does that and removes the TODO, explaining in the commit message that it's not related to 737 after all.

If you'd like to go further in a follow-up commit (which I think is not a high priority), I'd have some feedback on that direction:
- Take care to include `skipOffstage: false` where necessary when checking for something on a route that's not the top of the stack.
- I intentionally put `account1Route` (the one to be removed) under `account2Route` in the stack, to guard against buggy implementations that might use something like `NavigatorState.popUntil` (which starts from the top and stops at the first route that shouldn't be removed, even if there are ones that should be removed under it). This would be a nice feature to keep, but I guess not essential.

PIG208 replied:
Thanks for the detailed explanation! This has become #1257, but probably the refactor is not going to be a blocker for #737 so we can hold off reviewing that for now.

chrisbobbe reviewed (changes requested):
Thanks! Comments below.

PIG208 commented on lib/widgets/store.dart:
```diff
@@ -228,7 +235,16 @@ class _PerAccountStoreWidgetState extends State<PerAccountStoreWidget> {
+          (_) {
+            final navigator = Navigator.of(context);
+            navigator.removeRoute(widget.routeToRemoveOnLogout!);
+            if (!navigator.canPop()) {
+              // This ensures that the navigator stack is non-empty after the
+              // removal of the route.
+              unawaited(navigator.push(
+                MaterialWidgetRoute(page: const ChooseAccountPage())));
+            }
```
Yeah, a check for [both `isFirst` and `isCurrent`](https://api.flutter.dev/flutter/widgets/Route/isFirst.html) would be more appropriate.

PIG208 replied:
We now implement a `NavigatorObserver` based solution. It implements `didPop` and `didRemove` to push a choose-account page route proactively if it observes that the navigator stack is becoming empty.

PIG208 commented:
The PR has been updated to use a different approach maintaining the navigator stack. We are reusing a code path that throws `AccountNotFoundException` as a result of logging out the account. To ensure that the removal of a route does not end up with an empty navigator stack, we add a navigator observer to `ZulipApp` that pushes a choose-account page route when it becomes empty.

![image](https://github.com/user-attachments/assets/73b0e5c6-03ae-448c-832f-94960fd89685)

gnprice commented on lib/model/store.dart:
```diff
@@ -19,6 +19,7 @@ import '../api/backoff.dart';
 import '../api/route/realm.dart';
 import '../log.dart';
 import '../notifications/receive.dart';
+import '../widgets/actions.dart';
```
Prompted to look at this by https://github.com/zulip/zulip-flutter/pull/1257#discussion_r1917744040 . As I said there just now, let's avoid importing widgets/ from models/.

As the lib/widgets/actions.dart file's dartdoc says:
```dart
/// Methods that act through the Zulip API and show feedback in the UI.
///
/// The methods in this file can be thought of as higher-level wrappers for
/// some of the Zulip API endpoint binding methods in `lib/api/route/`.
/// But they don't belong in `lib/api/`, because they also interact with widgets
/// in order to present success or error feedback to the user through the UI.
```

the point of that file being in widgets/ is that its functions use widgets to give the user feedback. Since `logOutAccount` and its callee `unregisterToken` don't do that, they can move elsewhere.

An easy solution is to move those two functions to a file `lib/model/actions.dart`, i.e. under `model/`.

I think ultimately the right organization is probably that these belong as methods on GlobalStore, after all the more local-storage-focused methods that also act on accounts.

(Then in the future, as a matter of code organization, all those accounts-oriented methods' implementations will live on an AccountStore or AccountsStore, but the way one calls them will still be via GlobalStore — much like the existing ChannelStore and MessageStore etc. help keep PerAccountStore organized.)

PIG208 replied:
Echoing https://github.com/zulip/zulip-flutter/pull/1183#discussion_r1904045285, I assume the reason that we don't move it to `GlobalStore` right away is to prevent cluttering the class before we carry out the code reorganization you mentioned. The plan of moving it to `lib/model/actions.dart` sounds good to me (along with the `BuildContext` parameter to `GlobalStore` refactor).

PIG208 commented:
Updated the PR to resolve conflicts and move `logOutAccount` and `unregisterToken` to `lib/model/actions`. The relevant tests weren't moved. We can if we want. Those require a bit more care to be refactored from `testWidgets` tests to `test` tests, but it might not be worth it at the moment.

chrisbobbe commented on lib/model/store.dart:
```diff
@@ -912,7 +931,13 @@ class UpdateMachine {
         // at 1 kiB (at least on Android), and stack can be longer than that.
         assert(debugLog('Stack:\n$s'));
         assert(debugLog('Backing off, then will retry…'));
-        // TODO tell user if initial-fetch errors persist, or look non-transient
+        // TODO(#890): tell user if initial-fetch errors persist, or look non-transient
+        switch (e) {
+          case ZulipApiException(code: 'INVALID_API_KEY'):
+            // We cannot recover from this error through retrying.
+            // Leave it to [GlobalStore.loadPerAccount].
+            rethrow;
+        }
```
The log line 'Backing off, then will retry…' becomes inaccurate in some cases; can you adjust so that doesn't happen?

chrisbobbe commented on lib/model/store.dart:
```diff
@@ -912,7 +931,13 @@ class UpdateMachine {
-        // TODO tell user if initial-fetch errors persist, or look non-transient
+        // TODO(#890): tell user if initial-fetch errors persist, or look non-transient
+        switch (e) {
+          case ZulipApiException(code: 'INVALID_API_KEY'):
```
In manual testing on a dev server, where I changed an account's API key, the code I got was actually 'UNAUTHORIZED', not 'INVALID_API_KEY'. Have you seen 'INVALID_API_KEY' in manual testing? If you haven't done manual testing of the new functionality, please do that. 🙂

PIG208 replied:
Oh, I see! Didn't realize that my prior testing was limited to throwing `ZulipApiException` from certain places, but not with a dev server. Will do that. Thanks for pointing it out! :smiling_face_with_tear:

chrisbobbe reviewed (commented):
Thanks! Comments below, including one where I got something unexpected in manual testing.

PIG208 commented on lib/model/store.dart:
```diff
@@ -1044,7 +1069,13 @@ class UpdateMachine {
-      await _handlePollError(e);
+      try {
+        await _handlePollError(e);
+      } on AccountNotFoundException {
+        // Cannot recover by replacing the store because the account
+        // was logged out.
+        return;
+      }
```
Cannot `assert(_disposed)` here because the reload might not have happened if there was an error (likely during `_reloadPerAccount`) that prevents the store from being disposed.

I chose to not handle it in `_handlePollError` because of the `assert(_disposed)` that follows the `_handlePollError` call. This way we also maintain that `_handlePollError` has the store replaced. An alternative would be returning a `bool` from `_handlePollError`, indicating if the store gets disposed.

PIG208 replied:
Having tested this further, there are two cases `_handlePollError` should be aware of:

- The `_reloadPerAccount` call (`POST /register`) can fail due to an invalid API key. Handling this is necessary because otherwise it is a runtime error.
- The `Object error` parameter, i.e. the error from polling (`GET /events`) can be a `ZulipApiException` with HTTP status code 401, indicating the invalid API key error.
  Without handling it, it is marked as unexpected in the `default` branch under the `switch`, but we are still able to recover as long as we have error handling for the `_reloadPerAccount` call.

To handle the second case, we just need to acknowledge that such an error is expected and will be handled later.

PIG208 replied:
Let's defer the handling of `GET /events` failure to #1054.

PIG208 commented:
This is now stacked on top of #1340, which fixes a related issue that you may run into when invalidating the API key while on a message list page.

Also made some other changes apart from addressing the review comments on commit messages and code organization.

chrisbobbe commented on lib/model/store.dart:
```diff
@@ -149,7 +150,29 @@ abstract class GlobalStore extends ChangeNotifier {
+      store = await doLoadPerAccount(accountId);
+    } catch (e) {
+      switch (e) {
+        case HttpException(httpStatus: 401):
+          // The API key is invalid and the store can never be loaded
+          // unless the user retries manually.
+          final zulipLocalizations = GlobalLocalizations.zulipLocalizations;
+          reportErrorToUserModally(
+            zulipLocalizations.errorCouldNotConnectTitle,
+            details: zulipLocalizations.errorInvalidApiKeyMessage(
+              realmUrl.toString()));
+          // The account could be logged out from the choose-account page.
+          if (_accounts.containsKey(accountId)) {
+            await logOutAccount(this, accountId);
+          }
+          throw AccountNotFoundException();
```
Ooh actually: I think it would make sense to also make the  `reportErrorToUserModally` conditional on whether the account is still present, right?

Then, we don't give "could-not-authenticate" feedback about an account that has already been logged out. That feedback could be confusing or repetitive: confusing, if the account was logged out by the user's own action; repetitive, if the app chose to log out the account, because the app gives UI feedback when it does that.

Then also, no need to store `realmUrl` in a variable:

```dart
        case HttpException(httpStatus: 401):
          // The API key is invalid and the store can never be loaded
          // unless the user retries manually.
          final account = getAccount(accountId);
          if (account == null) {
            // The account was logged out during `await doLoadPerAccount`.
            // Here, that seems possible only by the user's own action;
            // the logout can't have been done programmatically.
            // Even if it were, it would have come with its own UI feedback.
            // Anyway, skip showing feedback, to not be confusing or repetitive.
            throw AccountNotFoundException();
          }
          final zulipLocalizations = GlobalLocalizations.zulipLocalizations;
          reportErrorToUserModally(
            zulipLocalizations.errorCouldNotConnectTitle,
            details: zulipLocalizations.errorInvalidApiKeyMessage(
              account.realmUrl.toString()));
          await logOutAccount(this, accountId);
          throw AccountNotFoundException();
```

chrisbobbe commented on lib/model/store.dart:
```diff
@@ -1218,8 +1248,17 @@ class UpdateMachine {
+    try {
+      await store._globalStore._reloadPerAccount(store.accountId);
+    } on AccountNotFoundException {
+      // The account was logged out while we retry to replace the store,
+      // but that's OK as long as other code will be removing it from the UI
+      // (usually by using [routeToRemoveOnLogout]).
+      assert(debugLog('… Event queue not replaced; account logged out.'));
```
nit: rather than "replace the store", how about "reload the store", as more directly what we mean, and matched with the name `_reloadPerAccount` in the code in the `try`.

PIG208 replied:
This is to match the other `debugLog` lines in the function, such as the ones that can lead to this line:
```dart
        assert(debugLog('Lost event queue for $store.  Replacing…'));
```

```dart
        assert(debugLog('BUG: Unexpected error in event polling: $error\n' // TODO(log)
          'Replacing event queue…'));
```

I think it is appropriate to use either "replace the event queue" or "reload the store" consistently. I feel that "replace the event queue" is less vague here because this attempted fix only makes sense when you relate it to the API layer.

chrisbobbe replied:
Ah yeah; "replace the queue" or "replace the event queue" would be fine for the code comment.

chrisbobbe commented on test/widgets/app_test.dart:
```diff
@@ -57,6 +59,99 @@ void main() {
+      final future = logOutAccount(testBinding.globalStore, eg.selfAccount.id);
+      await tester.pump(TestGlobalStore.removeAccountDuration);
+      await future;
+      check(pushedRoutes).single.isA<WidgetRoute>().page.isA<ChooseAccountPage>();
+      check(removedRoutes).single.isA<WidgetRoute>().page.isA<HomePage>();
```
Since a route is removed before a new one is pushed, how about ordering the `removedRoutes` check first? (Here and in other tests.)

chrisbobbe reviewed (commented):
Thanks! Comments below. I haven't yet tested the reload-on-dead-queue case manually yet; did you do that too?

PIG208 commented on lib/log.dart:
```diff
@@ -48,7 +48,7 @@ typedef ReportErrorCallback = void Function(String? message, {String? details});
 // from importing widget code, because the file is a dependency for the rest of
 // the app.
-ReportErrorCallback reportErrorToUserBriefly = defaultReportErrorToUserBriefly;
+ReportErrorCancellablyCallback reportErrorToUserBriefly = defaultReportErrorToUserBriefly;
```
Yeah. I think it was not clear without additional comments that this is describing a callback that accepts `null` for `message` when the error needs to be programmatically dismissed, instead of something UI specific.

chrisbobbe replied:
Ah, I think I understand now: `reportErrorToUserBriefly(null)` is a way to clear the snackbars programmatically.

PIG208 commented:
The PR has been updated. Thanks for the review!

> I haven't yet tested the reload-on-dead-queue case manually yet; did you do that too?

Do you mean the case when the event queue expires but the auth keys aren't valid? It works for me when I leave the app in the background after invalidating the API key.

chrisbobbe commented:
> Do you mean the case when the event queue expires but the auth keys aren't valid?

Yeah; (1) in your commit message:

```
The method loadPerAccount has two call sites, i.e. places
where we send register-queue requests:

1. _reloadPerAccount through [UpdateMachine._handlePollError]
2. perAccount through [PerAccountStoreWidget] (the common case)
```

Which I guess, thinking more about it, doesn't actually require the event queue expiring, right, because polling could fail for other reasons—including an invalid-API-key error :)

> It works for me when I leave the app in the background after invalidating the API key.

Great!

chrisbobbe commented:
> Yeah, a `GET /events` invalid-API-key error is typically a precursor to the invalid-API-key from the `_reloadPerAccount` code path, among other non-transient errors.

Yeah, that makes sense.

If `GET /events` has an invalid-API-key error, am I reading correctly that this code in `_handlePollError` is run?:

```dart
      default:
        assert(debugLog('BUG: Unexpected error in event polling: $error\n' // TODO(log)
          'Replacing event queue…'));
        _reportToUserErrorConnectingToServer(error);
        // Similar story to the _EventHandlingException case;
        // separate only so that that other case can print more context.
        // The bug here could be in the server if it's an ApiRequestException,
        // but our remedy is the same either way.
        isUnexpected = true;
```

If so, that seems wrong; this is something we expect to happen, though rarely, and the `_reportToUserErrorConnectingToServer` looks like redundant UI feedback.

PIG208 commented:
> am I reading correctly that this code in _handlePollError is run
Yes. I was thinking of deferring that to #1054 (https://github.com/zulip/zulip-flutter/pull/1183#discussion_r1947256331). For now, there is a TODO comment for this: 7334ccac31a7ea186af765a67530576854e9a2a4.

chrisbobbe commented:
Ah I see—thanks for reminding me of where you discussed this. :) I think deferring it to #1054 is fine.

chrisbobbe commented:
So, LGTM; marking for Greg's review.

gnprice commented on lib/model/store.dart:
```diff
@@ -1218,8 +1254,17 @@ class UpdateMachine {
+    try {
+      await store._globalStore._reloadPerAccount(store.accountId);
+    } on AccountNotFoundException {
+      // The account was logged out while we retry to replace the event queue,
+      // but that's OK as long as other code will be removing it from the UI
+      // (usually by using [routeToRemoveOnLogout]).
```
I don't understand how the second half of this sentence relates to the first. Why does that make it OK that the account was logged out — why would that be not-OK in the first place?

And what does "removing [the account] from the UI" mean?

… Hmm, it looks like this was copy-pasted from a comment in _PerAccountStoreWidgetState. But that one has a totally different meaning:
```
          // This widget will be showing [placeholder] perpetually,
          // but that's OK as long as other code will be removing it from the UI
          // (usually by using [routeToRemoveOnLogout]).
```

Showing a placeholder perpetually sure sounds like a bad thing — it's good to reason through why it'd be OK after all. And "removing [this widget] from the UI" makes sense because the widget is part of the UI in the first place.

PIG208 replied:
Yeah, I wanted to note that it is not bad when the event queue cannot be replaced. How about:

```
      // The event queue didn't get replaced because the account was logged out,
      // but that's OK as long as other code will ensure that the user
      // eventually navigates from the route associated with the old queue
      // (usually by using [routeToRemoveOnLogout]).
```

gnprice commented on lib/model/store.dart:
```diff
@@ -1218,8 +1254,17 @@ class UpdateMachine {
 
-    await store._globalStore._reloadPerAccount(store.accountId);
-    assert(_disposed);
+    try {
+      await store._globalStore._reloadPerAccount(store.accountId);
+    } on AccountNotFoundException {
```
Is this try/catch something we should have even before the new handling of auth errors?

I think it is: suppose the app loses its event queue and goes to reload initial data, and while it's doing that the user navigates to the choose-account page and logs out the account. Then I think `loadPerAccount` would hit its existing `throw AccountNotFoundException` line.

So this can be moved to a separate prep commit. I think doing that will also help simplify the explanation in the main commit's commit message.

PIG208 replied:
I found it pretty hard to hit this existing `throw AccountNotFoundException` line with `TestGlobalStore` or `LiveGlobalStore1, as we have an assertion at the start of `PerAccountStore.fromInitialSnapshot`:

```dart
    final account = globalStore.getAccount(accountId)!;
```

For both implementations, `PerAccountStore.fromInitialSnapshot` is called at the end with no asynchronous gap, meaning that the null-check will always fail ahead of `AccountNotFoundException` being raised, if the user logs out during the asynchronous gap from the /api/register-queue request or `globalStore.updateAccount`:

```dart
  static Future<UpdateMachine> load(GlobalStore globalStore, int accountId) async {
    Account account = globalStore.getAccount(accountId)!;
    final connection = globalStore.apiConnectionFromAccount(account);

    final stopwatch = Stopwatch()..start();
    final initialSnapshot = await _registerQueueWithRetry(connection);
    final t = (stopwatch..stop()).elapsed;
    assert(debugLog("initial fetch time: ${t.inMilliseconds}ms"));

    if (initialSnapshot.zulipVersion != account.zulipVersion
        || initialSnapshot.zulipMergeBase != account.zulipMergeBase
        || initialSnapshot.zulipFeatureLevel != account.zulipFeatureLevel) {
      account = await globalStore.updateAccount(accountId, AccountsCompanion(
        zulipVersion: Value(initialSnapshot.zulipVersion),
        zulipMergeBase: Value(initialSnapshot.zulipMergeBase),
        zulipFeatureLevel: Value(initialSnapshot.zulipFeatureLevel),
      ));
      connection.zulipFeatureLevel = initialSnapshot.zulipFeatureLevel;
    }

    final store = PerAccountStore.fromInitialSnapshot(
      globalStore: globalStore,
      accountId: accountId,
      connection: connection,
      initialSnapshot: initialSnapshot,
    );
    final updateMachine = UpdateMachine.fromInitialSnapshot(
      store: store, initialSnapshot: initialSnapshot);
    updateMachine.poll();
    if (initialSnapshot.serverEmojiDataUrl != null) {
      // TODO(server-6): If the server is ancient, just skip trying to have
      //   a list of its emoji.  (The old servers that don't provide
      //   serverEmojiDataUrl are already unsupported at time of writing.)
      unawaited(updateMachine.fetchEmojiData(initialSnapshot.serverEmojiDataUrl!));
    }
    // TODO do registerNotificationToken before registerQueue:
    //   https://github.com/zulip/zulip-flutter/pull/325#discussion_r1365982807
    unawaited(updateMachine.registerNotificationToken());
    return updateMachine;
  }
```

I think this is a bug that the `if (!_accounts.containsKey(accountId))` check was meant to catch. The error should be thrown earlier, ideally every time before the account is accessed, similar to the `if (mounted)` checks and `BuildContext`.

Modulo the bug, it would be helpful to have the try/catch here, except that it is unreachable until later.

chrisbobbe replied:
> Opened https://github.com/zulip/zulip-flutter/pull/1183 for this.

Did you mean a different issue/PR number here? That number refers to the current PR.

PIG208 replied:
Ah. It should be #1354.

gnprice replied:
Cool, thanks — as discussed on #1354, I agree that there should be a check earlier that throws AccountNotFoundException.

For exercising this case in a test (before the main changes at the tip of this PR), see new comment below: https://github.com/zulip/zulip-flutter/pull/1183#discussion_r1955499803

gnprice commented on lib/model/store.dart:
```diff
@@ -149,7 +150,34 @@ abstract class GlobalStore extends ChangeNotifier {
   /// and/or [perAccountSync].
   Future<PerAccountStore> loadPerAccount(int accountId) async {
```
nit in commit message:

    The method loadPerAccount has two call sites, i.e. places
    where we send register-queue requests:
    
    1. _reloadPerAccount through [UpdateMachine._handlePollError]
    2. perAccount through [PerAccountStoreWidget] (the common case)

I wouldn't call either of these "the common case" (which implies all other cases are rare or edge cases) — I think both of these are common. When I open the app it quite often was already in the background, but had been for more than 10 minutes, and that hits case 1.

gnprice commented on lib/model/store.dart:
```diff
@@ -149,7 +150,34 @@ abstract class GlobalStore extends ChangeNotifier {
+      store = await doLoadPerAccount(accountId);
+    } catch (e) {
+      switch (e) {
+        case HttpException(httpStatus: 401):
+          // The API key is invalid and the store can never be loaded
+          // unless the user retries manually.
```
Let's raise a thread on `#api design` asking for feedback on this condition `httpStatus: 401`. It seems quite reasonable, but it'd be good to make sure that everyone's on the same page about the semantics we'll be counting on from the server.

(For most things in the API we just work from the API docs with no further consultation — but this is an area of the API that's not clearly documented. The closest is https://chat.zulip.org/api/rest-error-handling , but for most errors that doesn't specify HTTP status codes, and indeed it specifies a Zulip API error code of INVALID_API_KEY which we learned at https://github.com/zulip/zulip-flutter/pull/1183#discussion_r1945865983 isn't what the server actually produces in the most obvious case of an API key being invalid.)

PIG208 replied:
[CZO conversation](https://chat.zulip.org/#narrow/channel/378-api-design/topic/general.20handling.20HTTP.20status.20code.20401)

From that thread, we confirmed that checking for 401 is the desired approach.

gnprice reviewed (commented):
Thanks @PIG208 for building this, and @chrisbobbe for the previous reviews! Comments below. I didn't finish reading the last commit's tests, but read everything else.

PIG208 commented on lib/widgets/app.dart:
```diff
@@ -211,7 +223,13 @@ class _ZulipAppState extends State<ZulipApp> with WidgetsBindingObserver {
+        navigatorObservers: [
+          if (widget.navigatorObservers != null)
+            ...widget.navigatorObservers!,
+            // This must be the last item to maintain the invariant
+            // that the navigator stack is always non-empty.
+            _EmptyStackNavigatorObserver(),
```
Leaving this comment out because we don't have other observers right now and that there's currently no issue prompting us to investigate further.

PIG208 commented on test/widgets/app_test.dart:
```diff
@@ -57,6 +59,99 @@ void main() {
+      check(pushedRoutes).isEmpty();
+    });
+
+    testWidgets('push route when popping last route on stack', (tester) async {
```
It does get popped (a dialog route is popped when dismissed), but we didn't track `poppedRoutes`. Adding it to the tests so that we have that knowledge as well.

gnprice replied:
Great, that clears it up, thanks.

gnprice reviewed (commented):
Thanks for the revision! Still reading, but posting comments now, since I see you've been replying to my related comments over on #1354.

gnprice commented on test/model/test_store.dart:
```diff
@@ -158,6 +163,9 @@ class TestGlobalStore extends GlobalStore {
     );
     updateMachines[accountId] = UpdateMachine.fromInitialSnapshot(
       store: store, initialSnapshot: initialSnapshot);
+    if (loadPerAccountPostDuration != null) {
+      await Future<void>.delayed(loadPerAccountPostDuration!);
+    }
     return Future.value(store);
```
Hypothetically, this can happen when `removeAccount` was called
    during `doLoadPerAccount`, possibly due to the user logging out
    from choose-account page while the store is getting reloaded.
    However, it is currently not reachable because of a bug in
    `UpdateMachine.load`.  See #1354.

We already have a test that reaches the analogous situation when initially loading an account: https://github.com/zulip/zulip-flutter/issues/1354#issuecomment-2658121787

So we should be able to use the same technique to test this change, without adding a new complication to TestGlobalStore. You'll see in that that existing test case that it uses the subclass LoadingTestGlobalStore, which gives the test much more direct hands-on control of how `doLoadPerAccount` behaves.

PIG208 replied:
Thanks! That's really helpful. The navigator tests are kept as-is, because they use `ZulipApp`, but the more bare-bone tests in `test/model/store_test.dart` are all switched to use this technique.

PIG208 commented on lib/widgets/app.dart:
```diff
@@ -234,6 +252,39 @@ class _ZulipAppState extends State<ZulipApp> with WidgetsBindingObserver {
+  void didRemove(Route<void> route, Route<void>? previousRoute) async {
+    if (previousRoute == null) {
+      // The route removed is the bottom-most one.
+      _pushRouteIfEmptyStack();
```
There is no distinction because I think there is no difference in user-facing behavior. Both implementation will be able to tell empty and non-empty stack apart correctly.

gnprice replied:
If there's no difference in behavior, then that means we can use either one of the two checks instead of using both of them, right?

In that case it'd be better to use just the `previousRoute == null` check — that one is a lot simpler.

PIG208 replied:
Oh I see! Sorry for the confusion. I was comparing an implementation with just `_pushRouteIfEmptyStack` and one with both `_pushRouteIfEmptyStack` and `previousRoute == null`. Both implementations would work.

Just having `_pushRouteIfEmptyStack` gives us the correct behavior, but only having `previousRoute == null` is insufficient because the bottom-most route can be removed while there are other routes on top of it.

gnprice commented on test/model/store_test.dart:
```diff
@@ -7,11 +7,13 @@ import 'package:flutter/foundation.dart';
 import 'package:zulip/api/model/model.dart';
 import 'package:zulip/api/route/events.dart';
 import 'package:zulip/api/route/messages.dart';
 import 'package:zulip/api/route/realm.dart';
+import 'package:zulip/model/actions.dart';
 import 'package:zulip/log.dart';
 import 'package:zulip/model/store.dart';
```
nit: keep imports sorted

gnprice commented on test/model/store_test.dart:
```diff
@@ -120,6 +122,39 @@ void main() {
+    addTearDown(testBinding.reset);
+
+    await testBinding.globalStore.insertAccount(eg.selfAccount.toCompanion(false));
+    testBinding.globalStore.loadPerAccountException = ZulipApiException(
+      routeName: '/register', code: 'UNAUTHORIZED', httpStatus: 401,
+      data: {}, message: '');
```
Hmm, seems like there ends up being a lot of copies of producing this same exception object. It'd be good to have a shared helper to encode the details — to help with making them consistently realistic, as well as making the test cases simpler to read.

I'll write a quick commit to do that, and push it to the branch or perhaps as a separate small PR.

gnprice replied:
Sent as #1355.

gnprice commented on test/widgets/app_test.dart:
```diff
@@ -57,6 +59,106 @@ void main() {
+      await tester.tap(find.byWidget(checkErrorDialog(tester,
+        expectedTitle: 'Could not connect',
+        expectedMessage:
+          'Your account at https://chat.example/ could not be authenticated.'
```
nit:
```suggestion
          'Your account at ${eg.selfAccount.realmUrl} could not be authenticated.'
```

otherwise this test case depends on the details of what realm URL eg.selfAccount happens to have

gnprice commented on test/widgets/app_test.dart:
```diff
@@ -57,6 +59,106 @@ void main() {
+      await testBinding.globalStore.insertAccount(eg.selfAccount.toCompanion(false));
+      await prepare(tester);
+
+      await tester.pump(kTryAnotherAccountWaitPeriod);
```
These test cases make sense, but they have enough complexity to them that I think they'd benefit from a line or two of comments narrating what each of these stanzas is doing. Would you try writing those?

PIG208 replied:
Added some comments. Mostly structured like "The navigator stack should contain…" with an optional explanation on what is expected to happen.

Because I feel like tracking the stack is essential for this kind of tests, and the `check(…Routes)` are useful checks, but not super intuitive for the reader.

gnprice commented on test/widgets/app_test.dart:
```diff
@@ -57,6 +59,106 @@ void main() {
+      pushedRoutes.clear();
+    }
+
+    testWidgets('do not push route to non-empty navigator stack', (tester) async {
```
nit: I think the narrative would be clearer if these three test cases:

```dart
    testWidgets('do not push route to non-empty navigator stack', (tester) async {
    // …
    testWidgets('push route when popping last route on stack', (tester) async {
    // …
    testWidgets('push route when removing last route on stack', (tester) async {
```

appeared in the opposite order.

In particular the "non-empty stack" case is a variation on the "popping last route" case, with some extra wrinkles added. So it's clearest to understand when taking that other case as a point of departure.

gnprice reviewed (commented):
OK, and done reading the new revision.

In addition to the new round of comments starting at https://github.com/zulip/zulip-flutter/pull/1183#pullrequestreview-2616610466 , a couple of subthreads continue on the previous round:
https://github.com/zulip/zulip-flutter/pull/1183#discussion_r1955469383
https://github.com/zulip/zulip-flutter/pull/1183#discussion_r1955491660

PIG208 commented:
Thanks for the review! I have updated the PR and stacked it atop #1355 (with the analyzer error fixed).

gnprice commented on test/model/store_test.dart:
```diff
@@ -970,6 +994,68 @@ void main() {
+          check(completers()).single.isCompleted.isFalse();
+
+          await completeLoading(async);
+          check(store).isLoading.isTrue();
```
This isLoading check and the one after it don't really check anything — `store` is still the original store, and would remain that way even if the reload succeeded.

gnprice commented on test/stdlib_checks.dart:
```diff
@@ -144,3 +145,7 @@ extension HttpMultipartFileChecks on Subject<http.MultipartFile> {
   Subject<String> get asString => has((x) => x.toString(), 'toString'); // TODO(checks): what's a good convention for this?
 }
+
+extension CompleterChecks<T> on Subject<Completer<T>> {
```
nit: put this above the JSON and URL/HTTP stuff — it's about futures, which are a core part of the language whereas those are more peripheral

gnprice commented on test/model/store_test.dart:
```diff
@@ -970,6 +994,68 @@ void main() {
+              eg.store(globalStore: globalStore, account: eg.selfAccount))
+          ..clear();
+        await future;
+        updateFromGlobalStore();
+        updateMachine.debugPauseLoop();
+        updateMachine.poll();
```
I think this group of tests would be easier to understand if it didn't try to use the system of shared locals that `updateFromGlobalStore` and the outer `preparePoll` are part of.

There's always a risk with that sort of setup that it becomes necessary to trace through a tangle of different layers to see what's going on. Here, these tests are using those variables but then don't actually want the whole package of how they're set up, so they also modify other aspects of how that system works, and it starts requiring a lot of flipping back and forth to understand.

Instead, this group can have its own local `globalStore`. It doesn't even need the other shared locals — `updateMachine` can be within this "prepare" helper, which can prepare the reload (the next few lines of `checkReloadFailure`) as well as the polling loop.

gnprice commented on test/model/store_test.dart:
```diff
@@ -1070,10 +1156,13 @@ class LoadingTestGlobalStore extends TestGlobalStore {
   Map<int, List<Completer<PerAccountStore>>> completers = {};
 
   @override
-  Future<PerAccountStore> doLoadPerAccount(int accountId) {
+  Future<PerAccountStore> doLoadPerAccount(int accountId) async {
     final completer = Completer<PerAccountStore>();
     (completers[accountId] ??= []).add(completer);
-    return completer.future;
+    final store = await completer.future;
+    updateMachines[accountId] = UpdateMachine.fromInitialSnapshot(
+      store: store, initialSnapshot: eg.initialSnapshot());
+    return store;
```
This is a bit unfortunate in that it complicates the picture for the existing tests that use this class. I think it's unnecessary if the new tests above just let go completely of the outer group's `preparePoll` and friends (as they've already half-done), as discussed in my preceding comment.

gnprice reviewed (commented):
Thanks for the revision!

I haven't yet gotten through the revised tests on the last commit, but here's comments on the rest.

PIG208 commented:
Thanks for the review! I have updated the PR and stacked it atop #1359 to make use of `apiException`.

gnprice commented on test/example_data.dart:
```diff
@@ -18,10 +19,52 @@ void _checkPositive(int? value, String description) {
+///
+/// https://zulip.com/api/get-events#bad_event_queue_id-errors
+ZulipApiException apiExceptionBadEventQueueId({
+  String queueId = 'fb67bf8a-c031-47cc-84cf-ed80accacda8',
```
Is this what a queue ID looks like? It differs from what we have in `eg.initialSnapshot`:
```dart
    queueId: queueId ?? '1:2345',
```

PIG208 replied:
Yeah. This comes from [/api/get-events](https://zulip.com/api/get-events#bad_event_queue_id-errors). We also use "abc:123" in another test. It might be good to have such a realistic value in test/example_data.dart to share.

gnprice replied:
I see, and in an actual /api/register-queue response I see a similar value.

I guess it must have changed at some point. Pretty sure in the past (years ago) it had the other form.

gnprice commented on test/model/store_test.dart:
```diff
@@ -757,11 +781,7 @@ void main() {
     }
 
     void prepareExpiredEventQueue() {
-      connection.prepare(httpStatus: 400, json: {
-        'result': 'error', 'code': 'BAD_EVENT_QUEUE_ID',
-        'queue_id': updateMachine.queueId,
-        'msg': 'Bad event queue ID: ${updateMachine.queueId}',
-      });
+      connection.prepare(apiException: eg.apiExceptionBadEventQueueId());
```
nit: Seems like this is more realistic when it uses the update machine's actual queue ID.

gnprice commented on test/model/store_test.dart:
```diff
@@ -972,6 +992,74 @@ void main() {
+    Future<void> prepareReload(FakeAsync async) async {
+      globalStore = LoadingTestGlobalStore(accounts: [eg.selfAccount]);
+      final future = globalStore.perAccount(eg.selfAccount.id);
+      completers()
+        ..single.complete(
+            eg.store(globalStore: globalStore, account: eg.selfAccount))
+        ..clear();
+      final store = await future;
```
nit: I think this is a bit clearer if it's made explicit that the `store` we end up with is the same one created with that `eg.store` call:
```suggestion
      final store = eg.store(globalStore: globalStore, account: eg.selfAccount);
      completers().single.complete(store);
      await future;
      completers().clear();
```

gnprice commented on test/model/store_test.dart:
```diff
@@ -972,6 +992,74 @@ void main() {
+      required FutureOr<void> Function() completeLoading,
+    }) {
+      awaitFakeAsync((async) async {
+        await prepareReload(async);
+
+        check(completers()).single.isCompleted.isFalse();
+        await completeLoading();
+        check(completers()).single.isCompleted.isTrue();
```
nit:
```suggestion
        await prepareReload(async);
        check(completers()).single.isCompleted.isFalse();

        await completeLoading();
        check(completers()).single.isCompleted.isTrue();
```

that way the rhythm of prep, check; then prep, check; is reflected visually in the stanzas between blank lines

gnprice commented on test/model/store_test.dart:
```diff
@@ -972,6 +992,74 @@ void main() {
+        check(globalStore.perAccountSync(eg.selfAccount.id)).isNull();
+
+        async.flushTimers();
+        // Reload never succeeds and there is no unhandled errors.
```
nit: "errors" is plural, so "there are":
```suggestion
        // Reload never succeeds and there are no unhandled errors.
```

gnprice reviewed (commented):
Thanks! Comments below on the parts that changed in the latest revision, all small.

Looking now at the last commit's tests, per https://github.com/zulip/zulip-flutter/pull/1183#pullrequestreview-2619186063 .

gnprice commented:
OK, finished reading. All looks good except (the comments above and) I felt the _PreventEmptyStack tests could use a bit more narrative beyond what you added at https://github.com/zulip/zulip-flutter/pull/1183#discussion_r1956829014 ; so I've pushed an additional to-be-squashed commit with some edits there. Please take a look and then go ahead and squash that in.

PIG208 commented:
Thanks for the reviews and edition! Updated the PR. For https://github.com/zulip/zulip-flutter/pull/1183#discussion_r1960629449, we can probably do a small sweep to factor out a shared example value, but that can be a follow-up.

gnprice commented:
Thanks for the revision! All LGTM except I guess #1355 and #1359 should get merged first.

chrisbobbe commented:
I've just merged those :)

PIG208 commented:
Thanks! Rebased the PR.

gnprice commented:
Great — merging.
